src/realm/util/file.cpp

```diff
@@ -1597,37 +1640,57 @@ std::optional<File::UniqueID> File::get_unique_id(const std::string& path)
         throw SystemError(GetLastError(), "CreateFileW failed");
     }
 
-    return get_unique_id(fileHandle);
+    return get_unique_id(fileHandle, path);
 #else // POSIX version
     struct stat statbuf;
     if (::stat(path.c_str(), &statbuf) == 0) {
+        if (statbuf.st_size == 0) {
+            // On exFAT systems the inode and device are not populated
+            // until the file has been allocated some space. This has led
+            // to bugs where a unique id returned here was reused by different
+            // files. Returning `none` here assumes that the caller knows this
+            // and will create non-empty files.
+            return none;
+        }
         return File::UniqueID(statbuf.st_dev, statbuf.st_ino);
     }
     int err = errno; // Eliminate any risk of clobbering
     // File doesn't exist
     if (err == ENOENT)
         return none;
-    throw SystemError(err, format_errno("fstat() failed: %1", err));
+    throw SystemError(err, format_errno("fstat() failed: %1 for '%2'", err, path));
 #endif
 }
 
-File::UniqueID File::get_unique_id(FileDesc file)
+File::UniqueID File::get_unique_id(FileDesc file, const std::string& debug_path)
 {
 #ifdef _WIN32 // Windows version
     REALM_ASSERT(file != nullptr);
     File::UniqueID ret;
     if (GetFileInformationByHandleEx(file, FileIdInfo, &ret.id_info, sizeof(ret.id_info)) == 0) {
-        throw std::system_error(GetLastError(), std::system_category(), "GetFileInformationByHandleEx() failed");
+        throw std::system_error(GetLastError(), std::system_category(),
+                                util::format("GetFileInformationByHandleEx() failed for '%1'", debug_path));
     }
 
     return ret;
 #else // POSIX version
     REALM_ASSERT(file >= 0);
     struct stat statbuf;
     if (::fstat(file, &statbuf) == 0) {
+        // On exFAT systems the inode and device are not populated
+        // until the file has been allocated some space. This has led
+        // to bugs where a unique id returned here was reused by different
+        // files. The following check ensures that this is not happening
```

Comment:
That is not fully true from what i understand. For empty file the uid can be anything, including the one already assigned to something else. This also happens when file is truncated and resized from 0 again. Do you think this is not an issue for encrypted mappings? The major problem how i can see it is that we make call for file uid on many operations in different times. There is no way currently to tell if something unexpected has happened in between. Ideally, i think, we should get it once for the proper path, and simply reuse the value for mappings and other operations.

Reply:
I updated the comment, please check if it is more accurate. Once a Realm is created, we never truncate to zero again and the cached unique id check I added will enforce this for future code.